File: src/fetch-elevation.js
/*
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { calculateWarnings } from "./gpx-parser.js";

/**
 * Fetches elevations for coordinate pairs from Open-Meteo with Open-Elevation fallback.
 * Processes in batches of 20 to avoid URL length constraints.
 * @param {Array} coords Array of [lat, lon] pairs
 * @param {Function} progressCallback Callback for progress reporting (current, total)
 * @returns {Promise<Array>} Array of elevations in meters
 */
async function fetchElevationsAPI(coords, progressCallback) {
  const batchSize = 20;
  const elevations = [];

  for (let i = 0; i < coords.length; i += batchSize) {
    const batch = coords.slice(i, i + batchSize);
    if (progressCallback) {
      progressCallback(Math.min(i + batchSize, coords.length), coords.length);
    }

    let success = false;
    let retries = 3;
    let delayMs = 1000;

    while (!success && retries > 0) {
      try {
        // Try Open-Meteo
        const lats = batch.map((c) => c[0]).join(",");
        const lons = batch.map((c) => c[1]).join(",");
        const url = `https://api.open-meteo.com/v1/elevation?latitude=${lats}&longitude=${lons}`;

        const response = await fetch(url);
        if (!response.ok) throw new Error(`HTTP Error ${response.status}`);
        const data = await response.json();
        
        if (data && data.elevation) {
          elevations.push(...data.elevation);
          success = true;
        } else {
          throw new Error("No elevation data in Open-Meteo response");
        }
      } catch (err) {
        console.warn(`Open-Meteo batch failed (retries remaining: ${retries - 1}):`, err);
        
        // Try Open-Elevation fallback
        try {
          const url = "https://api.open-elevation.com/api/v1/lookup";
          const payload = {
            locations: batch.map((c) => ({ latitude: c[0], longitude: c[1] })),
          };

          const response = await fetch(url, {
            method: "POST",
            headers: {
              "Content-Type": "application/json",
            },
            body: JSON.stringify(payload),
          });
          if (!response.ok) throw new Error(`HTTP Error ${response.status}`);
          const data = await response.json();

          if (data && data.results) {
            elevations.push(...data.results.map((loc) => loc.elevation));
            success = true;
          } else {
            throw new Error("No elevation data in Open-Elevation response");
          }
        } catch (err2) {
          console.warn("Open-Elevation fallback also failed:", err2);
          retries--;
          if (retries > 0) {
            await new Promise((resolve) => setTimeout(resolve, delayMs));
            delayMs *= 2; // exponential backoff
          }
        }
      }
    }

    if (!success) {
      console.error(`Failed to fetch elevations for batch starting at index ${i}. Filling with 0.0.`);
      elevations.push(...Array(batch.length).fill(0.0));
    }

    // Rate limit safeguard sleep
    await new Promise((resolve) => setTimeout(resolve, 300));
  }

  return elevations;
}

/**
 * Corrects and interpolates elevations for an in-memory route object.
 * @param {Object} route The parsed route object to modify
 * @param {Function} progressCallback Callback to display progress
 * @returns {Promise<void>}
 */
export async function correctRouteElevations(route, progressCallback) {
  const trackpoints = route.trackpoints;
  const N = trackpoints.length;
  if (N === 0) return;

  const coords = trackpoints.map((pt) => [pt.lat, pt.lon]);

  // Sample every 10th point and always include the last point (N-1)
  const step = 10;
  const sampleIndices = [];
  for (let i = 0; i < N; i += step) {
    sampleIndices.push(i);
  }
  if (sampleIndices[sampleIndices.length - 1] !== N - 1) {
    sampleIndices.push(N - 1);
  }

  const sampleCoords = sampleIndices.map((idx) => coords[idx]);

  // Fetch sampled elevations
  const sampleElevations = await fetchElevationsAPI(sampleCoords, progressCallback);

  // Map back and linearly interpolate
  const fullElevations = Array(N).fill(0.0);
  sampleIndices.forEach((sampleIdx, k) => {
    fullElevations[sampleIdx] = sampleElevations[k];
  });

  for (let k = 0; k < sampleIndices.length - 1; k++) {
    const startIdx = sampleIndices[k];
    const endIdx = sampleIndices[k + 1];
    const startEle = fullElevations[startIdx];
    const endEle = fullElevations[endIdx];

    const diff = endIdx - startIdx;
    if (diff > 1) {
      for (let i = startIdx + 1; i < endIdx; i++) {
        const fraction = (i - startIdx) / diff;
        fullElevations[i] = startEle + fraction * (endEle - startEle);
      }
    }
  }

  // Write elevations back to trackpoints and recalculate grades
  let totalElevationGain = 0;
  let totalElevationLoss = 0;

  trackpoints.forEach((pt, i) => {
    pt.ele = fullElevations[i];

    if (i > 0) {
      const prev = trackpoints[i - 1];
      const segmentDist = pt.dist_m - prev.dist_m;
      const eleDiff = pt.ele - prev.ele;

      if (eleDiff > 0) {
        totalElevationGain += eleDiff;
      } else {
        totalElevationLoss += Math.abs(eleDiff);
      }

      if (segmentDist > 0.1) {
        pt.grade = (eleDiff / segmentDist) * 100;
      } else {
        pt.grade = 0;
      }
    } else {
      pt.grade = 0;
    }
  });

  route.totalElevationGain = totalElevationGain;
  route.totalElevationLoss = totalElevationLoss;

  // Sync waypoint elevations to closest trackpoint elevations
  route.waypoints.forEach((wpt) => {
    if (wpt.closestTrackpointIndex !== undefined) {
      wpt.ele = trackpoints[wpt.closestTrackpointIndex].ele;
    }
  });

  // Recalculate warnings (climb calculations depend on elevations)
  calculateWarnings(route);
}

File: fetch_elevation.py
# Copyright 2026 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import xml.etree.ElementTree as ET
import urllib.request
import json
import urllib.parse
import os
import time

def fetch_from_open_meteo(batch):
    lats = ",".join(str(c[0]) for c in batch)
    lons = ",".join(str(c[1]) for c in batch)
    url = f"https://api.open-meteo.com/v1/elevation?latitude={lats}&longitude={lons}"
    
    req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0'})
    with urllib.request.urlopen(req, timeout=10) as response:
        data = json.loads(response.read().decode())
        if 'elevation' in data:
            return data['elevation']
    raise Exception("No elevation data in Open-Meteo response")

def fetch_from_open_elevation(batch):
    url = "https://api.open-elevation.com/api/v1/lookup"
    payload = {
        "locations": [{"latitude": c[0], "longitude": c[1]} for c in batch]
    }
    req_data = json.dumps(payload).encode('utf-8')
    req = urllib.request.Request(
        url, 
        data=req_data, 
        headers={'Content-Type': 'application/json', 'User-Agent': 'Mozilla/5.0'}
    )
    with urllib.request.urlopen(req, timeout=10) as response:
        data = json.loads(response.read().decode())
        if 'results' in data:
            return [loc['elevation'] for loc in data['results']]
    raise Exception("No elevation data in Open-Elevation response")

def fetch_elevations(coords):
    # Conservative batch size of 20 to prevent HTTP 414 and 400.
    batch_size = 20
    elevations = []
    
    for i in range(0, len(coords), batch_size):
        batch = coords[i:i+batch_size]
        print(f"Fetching batch {i // batch_size + 1}/{len(coords) // batch_size + 1} ({len(batch)} points)...", flush=True)
        
        success = False
        retries = 3
        backoff = 2
        
        while not success and retries > 0:
            # Method 1: Try Open-Meteo
            try:
                elevations.extend(fetch_from_open_meteo(batch))
                success = True
            except Exception as e:
                print(f"Open-Meteo failed: {e}. Trying fallback Open-Elevation...", flush=True)
                # Method 2: Try Open-Elevation fallback
                try:
                    elevations.extend(fetch_from_open_elevation(batch))
                    success = True
                except Exception as e2:
                    print(f"Open-Elevation fallback failed: {e2}.", flush=True)
                    retries -= 1
                    if retries > 0:
                        print(f"Retrying in {backoff} seconds...", flush=True)
                        time.sleep(backoff)
                        backoff *= 2
                        
        if not success:
            print("Failed to fetch batch from all providers. Filling with 0.0.", flush=True)
            elevations.extend([0.0] * len(batch))
            
        # 0.3 second sleep between requests to be gentle and prevent rate-limiting (429)
        time.sleep(0.3)
            
    return elevations

def process_gpx(filepath):
    print(f"Processing {filepath}...", flush=True)
    ET.register_namespace('', 'http://www.topografix.com/GPX/1/1')
    
    tree = ET.parse(filepath)
    root = tree.getroot()
    
    ns = {'gpx': 'http://www.topografix.com/GPX/1/1'}
    
    trkpts = root.findall('.//gpx:trkpt', ns)
    N = len(trkpts)
    print(f"Found {N} trackpoints.", flush=True)
    
    if not N:
        return
        
    coords = []
    for pt in trkpts:
        lat = float(pt.get('lat'))
        lon = float(pt.get('lon'))
        coords.append((lat, lon))
        
    # Sample every 10th point (~220 meters) and always include the last point (N-1)
    step = 10
    sample_indices = list(range(0, N, step))
    if (N - 1) not in sample_indices:
        sample_indices.append(N - 1)
        
    sample_coords = [coords[idx] for idx in sample_indices]
    print(f"Sampling {len(sample_coords)} points for elevation APIs (10x reduction)...", flush=True)
    
    sample_elevations = fetch_elevations(sample_coords)
    
    # Map back and linearly interpolate
    full_elevations = [0.0] * N
    for idx, ele in zip(sample_indices, sample_elevations):
        full_elevations[idx] = ele
        
    for k in range(len(sample_indices) - 1):
        start_idx = sample_indices[k]
        end_idx = sample_indices[k+1]
        start_ele = full_elevations[start_idx]
        end_ele = full_elevations[end_idx]
        
        diff = end_idx - start_idx
        if diff > 1:
            for i in range(start_idx + 1, end_idx):
                fraction = (i - start_idx) / diff
                full_elevations[i] = start_ele + fraction * (end_ele - start_ele)
                
    for pt, ele in zip(trkpts, full_elevations):
        ele_elem = pt.find('gpx:ele', ns)
        if ele_elem is not None:
            ele_elem.text = f"{ele:.2f}"
        else:
            new_ele = ET.Element('{http://www.topografix.com/GPX/1/1}ele')
            new_ele.text = f"{ele:.2f}"
            pt.insert(0, new_ele)
            
    tree.write(filepath, xml_declaration=True, encoding='utf-8')
    print(f"Finished writing elevations to {filepath}\n", flush=True)

if __name__ == "__main__":
    import sys
    if len(sys.argv) > 1:
        filepath = sys.argv[1]
        if os.path.exists(filepath):
            process_gpx(filepath)
        else:
            print(f"File not found: {filepath}", flush=True)
    else:
        print("Usage: python3 fetch_elevation.py <path_to_gpx>")
